sockeye/config.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not
+# use this file except in compliance with the License. A copy of the License
+# is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import copy
+
+import yaml
+
+
+class Config(yaml.YAMLObject):
+    """
+    Base configuration object that supports freezing of members and JSON (de-)serialization.
+    Actual Configuration should subclass this object.
+    """
+    yaml_tag = u'!Config'
+
+    def __init__(self):
+        self._frozen = False
+
+    def __setattr__(self, key, value):
+        if hasattr(self, '_frozen') and self._frozen:
+            raise AttributeError("Cannot set '%s' in frozen config" % key)
+        object.__setattr__(self, key, value)
+
+    def freeze(self):
+        """
+        Freezes this Config object, disallowing modification or addition of any parameters.
+        """
+        if hasattr(self, '_frozen') and self._frozen:  # It's ok to freeze an already frozen config
+            return
+        self._frozen = True
+        for k, v in self.__dict__.items():
```

Comment:
I saw that but didn't really get that would happen. You mean if a child config has a reference to the parent? Because just by going through __dict__.items() we won't call 'ourselves' here.